db/compaction/compaction_job.cc

```diff
@@ -290,6 +296,128 @@ void CompactionJob::Prepare() {
+         static_cast<uint64_t>(ranges_size_limit)});
+
+    TEST_SYNC_POINT_CALLBACK("CompactionJob::GetNumPlannedSubcompactions:Init",
+                             &num_planned_subcompactions);
+    if (max_subcompactions_limit < num_planned_subcompactions) {
+      // Step 2: Check if the initial number exceeds user-configured
+      // max_subcompaction. If yes, we need to compute the extra subcompaction
+      // based on compaction limits (the extra subcompactions should not
+      // be larger than that)
+      int extra_num_subcompactions_limit =
+          std::min((int)(num_planned_subcompactions - max_subcompactions_limit),
+                   available_bg_compactions_db_limit_);
+
+      TEST_SYNC_POINT_CALLBACK(
+          "CompactionJob::GetNumPlannedSubcompactions:"
+          "ExtraCompactionsAgainstDBLimits",
+          &extra_num_subcompactions_limit);
+
+      // Step 3: Try to reserve threads for extra subcompactions. The planned
+      // subcompactions may change again if we cannot obtain sufficient threads
+      extra_num_subcompaction_threads_reserved_ =
+          env_->ReserveThreads(extra_num_subcompactions_limit, thread_pri_);
+      num_planned_subcompactions =
+          static_cast<uint64_t>(extra_num_subcompaction_threads_reserved_) +
+          max_subcompactions_limit;
+      TEST_SYNC_POINT_CALLBACK(
+          "CompactionJob::GetNumPlannedSubcompactions:Finalize",
+          &num_planned_subcompactions);
+    }
+  }
+  return num_planned_subcompactions;
+}
+
+void CompactionJob::UpdateResourcesAndLimits(
+    uint64_t num_actual_subcompactions, uint64_t num_planned_subcompactions) {
+  // We only need to update resources and limits if the actual subcompactions
+  // is less than the planned one and we did reserve some resources earlier
+  if (extra_num_subcompaction_threads_reserved_ > 0 &&
+      num_actual_subcompactions < num_planned_subcompactions) {
+    // We cannot release threads more than what we reserved before
+    int additional_bg_compactions = env_->ReleaseThreads(
+        std::min((int)(num_planned_subcompactions - num_actual_subcompactions),
+                 extra_num_subcompaction_threads_reserved_),
+        thread_pri_);
+    // Update the number of reserved threads and the number of background
+    // scheduled compactions for this compaction job
+    extra_num_subcompaction_threads_reserved_ -= additional_bg_compactions;
+    db_mutex_->AssertHeld();
+    // Update bg_compaction_scheduled_ or bg_bottom_compaction_scheduled_
+    // depending on if this compaction has the bottommost priority
+    if (thread_pri_ == Env::Priority::BOTTOM && bottommost_level_) {
+      if (bg_bottom_compaction_scheduled_ != nullptr) {
+        *bg_bottom_compaction_scheduled_ +=
+            extra_num_subcompaction_threads_reserved_;
+      }
+    } else if (bg_compaction_scheduled_ != nullptr) {
+      *bg_compaction_scheduled_ += extra_num_subcompaction_threads_reserved_;
+    }
```

Comment:
Seems to me that this update is currently done only when `if (extra_num_subcompaction_threads_reserved_ > 0 &&
      num_actual_subcompactions < num_planned_subcompactions)`

Would it be possible that we will still need this update when `num_actual_subcompactions==num_planned_subcompactions`?

Reply:
I don't think we need this update when they are equal (indicating that we exactly reserved sufficient sources)/

Reply:
I am a bit confused. I thought `bg_compaction_scheduled_ / I bg_bottom_compaction_scheduled_` update is needed whenever we need to reserve threads from the pool, e.g, when `max_subcompactions_limit < num_planned_subcompactions`. And `num_actual_subcompactions==num_planned_subcompactions` only means the "mean/sum" trick indeed results in exactly the `num_planned_subcompactions`. It's not saying we don't need to reserve threads from the pool though.

Am I missing something??

Reply:
Another way to see my point: `bg_compaction_scheduled_ / bg_bottom_compaction_scheduled_` update is done in `ReleaseResourcesAndLimits()` without considering `num_actual_subcompactions < num_planned_subcompactions` this constraint at all. I will expect `bg_compaction_scheduled_ / bg_bottom_compaction_scheduled_` update to be symmetric across `Release` and `Reserve + Shrink if needed`

Reply:
My bad, and you are right. I will update this in the next version.

Reply:
I can't understand as the code link no longer works. Is this resolved?